services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowHistorySpec.groovy

```diff
@@ -0,0 +1,140 @@
+        Long timestampAfterCreate = System.currentTimeMillis() / 1000L
+        checkHistoryCreateAction(flow.id, timestampAfterCreate)
+
+        when: "Update the created flow"
+        flowHelper.updateFlow(flow.id, flow.tap { it.description = it.description + "updated" })
+
+        then: "History record is created after updating the flow"
+        Long timestampAfterUpdate = System.currentTimeMillis() / 1000L
+        checkHistoryUpdateAction(flow.id, timestampAfterUpdate)
+
+        when: "Delete the updated flow"
+        flowHelper.deleteFlow(flow.id)
+
+        then: "History is still available for the deleted flow"
+        checkHistoryDeleteAction(flow.id, timestampAfterUpdate)
+    }
+
+    def "History records are created for the create/update actions using default timeline"() {
+        when: "Create a flow"
+        def (Switch srcSwitch, Switch dstSwitch) = topology.activeSwitches
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flowHelper.addFlow(flow)
+
+        then: "History record is created"
+        checkHistoryCreateAction(flow.id)
+
+        when: "Update the created flow"
+        flowHelper.updateFlow(flow.id, flow.tap { it.description = it.description + "updated" })
+
+        then: "History record is created after updating the flow"
+        checkHistoryUpdateAction(flow.id)
+
+        when: "Delete the updated flow"
+        flowHelper.deleteFlow(flow.id)
+
+        then: "History is still available for the deleted flow"
+        checkHistoryDeleteAction(flow.id)
+    }
+
+    def "History should not be returned in case timeline is incorrect (timeBefore > timeAfter)"() {
+        when: "Create a flow"
+        def (Switch srcSwitch, Switch dstSwitch) = topology.activeSwitches
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flowHelper.addFlow(flow)
+
+        then: "History record is created"
+        Long timestampAfterCreate = System.currentTimeMillis() / 1000L
+        checkHistoryCreateAction(flow.id, timestampAfterCreate)
+
+        when: "Try to get history for incorrect timeline"
+        def flowH = northbound.getFlowHistory(flow.id, timestampAfterCreate, timestampBefore)
+
+        then: "History record is NOT returned"
+        flowH.isEmpty()
+
+        and: "Cleanup: restore default state(remove created flow)"
+        flowHelper.deleteFlow(flow.id)
+    }
+
+    def "History should not be returned in case flow was never created"() {
+        when: "Try to get history for incorrect flowId"
+        def flowHistory = northbound.getFlowHistory(NON_EXISTENT_FLOW_ID)
+
+        then: "History record is NOT returned"
+        flowHistory.isEmpty()
+    }
+
+    void checkHistoryCreateAction(flowId, timestampAfterCreate = null) {
+        def timeBefore = timestampAfterCreate ? timestampBefore : null
+        def flowHistory = northbound.getFlowHistory(flowId, timeBefore, timestampAfterCreate)
+
+        assert flowHistory.size() == 1
```

Comment:
I would have moved the above code (L101>L104) directly to test and leave only `assertHistoryCreate/Update` methods, currently looks like obfuscation to me. But I don't have very strong opinion here, @ylobankov thoughts?

Reply:
I am fine with the current implementation, but I think this movement will give us more readability.

Reply:
You voted with both hands to move it, 
so I will do it)